canonical/ubuntu-pro-client
# refactor: implementation-agnostic interface to readurl

## Why is this needed?

This PR moves all http-related code into its own module and defines a simple interface to the `readurl` function.

The previous interface to `readurl` had different return types depending on the request, and it propogated exceptions that were raised from the urllib library. Those urllib exceptions were then caught and used in various places throughout the codebase.

The tls-in-tls implementation will require swapping out the urllib request function for a pycurl-based one, which will not have any of the same return types or exceptions as urllib, so we need to stop using urllib-specific things everywhere.

To achieve this, I've defined a simple HTTPResponse tuple that will always be returned by the `readurl` function - this should be easily implemented by any underlying http implementation. Notice that the function always returns this tuple, even if there is a http code above 300. Previously we expected/excepted errors on 4xx/5xx responses, so I've changed a lot of the calling code to check the return code rather than except errors. The only error that is raised is now our `exceptions.UrlError` and I've redefined it to only be raised if something about the connection fails. I think that overall this interface is simpler and will be easier to implement than raising errors for 4xx/5xx http responses. LMK what you think

TODO:
- [ ] add above explanation to commit message
- [x] let CI run to catch integration test errors
- [ ] run several cloud pro tests to make sure auto-attach still works




## Test Steps

Run a variety of cloud/pro tests to make sure all the metadata requests are still working



## Checklist

 - [n/a] I have updated or added any documentation accordingly

## Does this PR require extra reviews?
